docs/core/v3.0.rst

```diff
@@ -0,0 +1,1761 @@
+    `j`, ...), perform ``get(data_key(P, j, i, ...), value)``. The returned
+    value is the serialisation of the corresponding chunk, encoded
+    according to the array metadata stored at ``array_meta_key(P)``.
+
+**Discover children of a group**
+
+    To discover the children of a group at hierarchy path `P`, perform
+    ``list_dir("meta/root" + P + "/")``. Any returned key ending in
+    ".array.json" indicates an array. Any returned key ending in
+    ".group.json" indicates a group. Any returned prefix indicates a
+    child group implied by some descendant.
+
+    For example, if a group is created at path "/foo/bar" and an array
+    is created at path "/foo/baz/qux", then the store will contain the
+    keys "meta/root/foo/bar.group.json" and
+    "meta/root/foo/bar/baz/qux.array.json". Groups at paths "/",
+    "/foo" and "/foo/baz" have not been explicitly created but are
+    implied by their descendants. To list the children of the group at
+    path "/foo", perform ``list_dir("meta/root/foo/")``, which will
+    return the key "meta/root/foo/bar.group.json" and the prefix
+    "meta/root/foo/baz/". From this it can be inferred that child
+    groups "/foo/bar" and "/foo/baz" are present.
+
+    If a store does not support any of the list operations then
+    discovery of group children is not possible, and the contents of
+    the hierarchy must be communicated by some other means, such as
+    via an extension, or via some out of band communication.
+
+**Discover all nodes in a hierarchy**
+
+    To discover all nodes in a hierarchy, one can call
+    ``list_prefix("meta/root/")``. All keys represent either explicit group or
+    arrays. All intermediate prefixes ending in a ``/`` are implicit
+    groups.
+
+**Erase a group or array**
+
+    To erase an array at path `P`:
+      - erase the metadata document for the array, ``erase(array_meta_key(P))``
+      - erase all data keys which prefix have path pointing to this array,
+        ``erase_prefix("data/root" + P + "/")``
+
+    To erase an implicit group at path `P`:
+      - erase all nodes under this group - it should be sufficient to
+        perform ``erase_prefix("meta/root" + P + "/")`` and
+        ``erase_prefix("data/root" + P + "/")``.
+
+    To erase an explicit group at path `P`:
+      - erase the metadata document for the group, ``erase(group_meta_key(P))``
+      - erase all nodes under this group - it should be sufficient to
+        perform ``erase_prefix("meta/root" + P + "/")`` and
+        ``erase_prefix("data/root" + P + "/")``.
+
+**Determine if a node exists**
+
+    To determine if a node exists at path ``P``, try in the following
+    order ``get(array_meta_key(P))`` (success implies an array at
+    ``P``); ``get(group_meta_key(P))`` (success implies an explicit
+    group at ``P``); ``list_dir("meta/root" + P + "/")`` (non-empty
+    result set implies an implicit group at ``P``).
```

Comment:
This feels a little awkward. I would think that the first operation one should when opening a new store is just "discover all notes in a hierarchy". So under what (other) circumstances would you want to "determine if a node exists"? Could we just remove this section?

Reply:
I think the use case would be "requests are slow" and/ or "there are many nodes" so you don't want to discover all the nodes if you already know what path to retrieve. Maybe I have a thousand images in a store, but only want one. So:

```python
P = "some/path
z = zarr.open("s3://bucket/root")  # reads root metadata zarr.json
P in z  # Performs the above example
```

I do think this is awkward though.

Reply:
Of course it is up to the implementation, but in my mind, listing the contents of a group/root would not be done by default, but only upon explicit request, except perhaps for an interactive user interface.  Additionally, new arrays or groups may be added after the zarr repo is first "opened", so just checking once initially may not be sufficient.

But I also think all of these operation descriptions, while perhaps helpful implementation aids, are not really critical parts of the spec and it may be better not to focus on them until after we've finalized the data format itself, since finalizing the data format is blocking usage of zarr v3.

Reply:
I added a TODO note for the operations section for now, mentioning that it is not finalized yet. See e5297ae of #171.